packages/runtime/container-runtime/src/batchTracker.ts

```diff
@@ -0,0 +1,82 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { assert } from "@fluidframework/common-utils";
+import { ISequencedDocumentMessage } from "@fluidframework/protocol-definitions";
+import { ChildLogger } from "@fluidframework/telemetry-utils";
+import EventEmitter from "events";
+
+export class BatchTracker {
+    private readonly logger: ITelemetryLogger;
+    private startBatchSequenceNumber: number | undefined;
+    private trackedBatchCount: number = 0;
+    private batchProcessingStartTimeStamp: number | undefined;
+
+    constructor(
+        private readonly batchEventEmitter: EventEmitter,
+        logger: ITelemetryLogger,
+        private readonly opCountThreshold: number,
+        private readonly batchCountSamplingRate: number,
+        private readonly dateTimeProvider: () => number = () => Date.now(),
+    ) {
+        this.logger = ChildLogger.create(logger, "Batching");
+
+        this.batchEventEmitter.on("batchBegin", (message: ISequencedDocumentMessage) => {
+            this.startBatchSequenceNumber = message.sequenceNumber;
+            this.batchProcessingStartTimeStamp = this.dateTimeProvider();
+            this.trackedBatchCount++;
+        });
+
+        this.batchEventEmitter.on("batchEnd", (error: any | undefined, message: ISequencedDocumentMessage) => {
+            assert(
+                this.startBatchSequenceNumber !== undefined && this.batchProcessingStartTimeStamp !== undefined,
+                "batchBegin must fire before batchEnd");
+
+            const opCount = message.sequenceNumber - this.startBatchSequenceNumber + 1;
+            if (opCount >= this.opCountThreshold) {
+                this.logger.sendErrorEvent({
+                    eventName: "TooManyOps",
+                    opCount,
+                    threshold: opCountThreshold,
+                    referenceSequenceNumber: message.referenceSequenceNumber,
```

Comment:
Here and below: I'm not sure a bunch of background info , like ref seq # gives us anything. That kind of info is useful when tracking errors through lifetime of a file, but for events that are for aggregated statistics it's just extra data that will never get used. I'd keep it lean, and add more when we need it.